crates/matrix-sdk/src/authentication/qrcode/rendezvous_channel.rs

```diff
@@ -0,0 +1,433 @@
+// Copyright 2024 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::time::Duration;
+
+use http::{
+    header::{CONTENT_TYPE, ETAG, EXPIRES, IF_MATCH, IF_NONE_MATCH, LAST_MODIFIED},
+    HeaderMap, HeaderName, Method, StatusCode,
+};
+use ruma::api::{
+    client::error::ErrorKind,
+    error::{FromHttpResponseError, HeaderDeserializationError, IntoHttpError, MatrixError},
+    EndpointError,
+};
+use tracing::{debug, instrument, trace, warn};
+use url::Url;
+
+use crate::{http_client::HttpClient, HttpError, RumaApiError};
+
+const TEXT_PLAIN_CONTENT_TYPE: &str = "text/plain";
+const RECEIVE_TIMEOUT: Duration = Duration::from_secs(1);
+
+type Etag = String;
+
+/// Get a header from a [`HeaderMap`] and parse it as a UTF-8 string.
+fn get_header(
+    header_map: &HeaderMap,
+    header_name: &HeaderName,
+) -> Result<String, FromHttpResponseError<RumaApiError>> {
+    let header = header_map
+        .get(header_name)
+        .ok_or(HeaderDeserializationError::MissingHeader(ETAG.to_string()))?;
+
+    let header = header.to_str()?.to_owned();
+
+    Ok(header)
+}
+
+/// The result of the [`RendezvousChannel::create_inbound()`] method.
+pub(super) struct InboundChannelCreationResult {
+    /// The connected [`RendezvousChannel`].
+    pub channel: RendezvousChannel,
+    /// The initial message we received when we connected to the
+    /// [`RendezvousChannel`].
+    ///
+    /// This is currently unused, but left in for completeness sake.
+    #[allow(dead_code)]
+    pub initial_message: Vec<u8>,
+}
+
+struct RendezvousGetResponse {
+    pub status_code: StatusCode,
+    pub etag: String,
+    // TODO: This is currently unused, but will be required once we implement the reciprocation of
```

Comment:
Can you create an issue in github.com/matrix-org/matrix-rust-sdk/ so that we don't forget about it please?

Reply:
I usually avoid doing this because [trouble](https://github.com/folke/trouble.nvim) is much better at showing/searching TODO items in the codebase than github is.

Another point is that some customer already requested this, so it is somewhere already present and will turn up on our plate.

That being said, I wouldn't mind adding it if the above didn't convince you that it isn't necessary.

Reply:
Fiiiine 😉.